spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/repository/query/SqlSpannerQuery.java

```diff
@@ -103,11 +117,93 @@ private String resolveEntityClassNames(String sql) {
 		return result;
 	}
 
+	private void resolveSpELTags(QueryTagValue queryTagValue) {
+		Expression[] expressions = detectExpressions(queryTagValue.sql);
+		StringBuilder sb = new StringBuilder();
+		Map<Object, String> valueToTag = new HashMap<>();
+		int tagNum = 0;
+		for (Expression expression : expressions) {
+			if (expression instanceof LiteralExpression) {
+				sb.append(expression.getValue(String.class));
+			}
+			else if (expression instanceof SpelExpression) {
+				Object value = expression.getValue(this.evaluationContextProvider
+						.getEvaluationContext(this.queryMethod.getParameters(),
+								queryTagValue.intialParams));
+				if (valueToTag.containsKey(value)) {
+					sb.append("@" + valueToTag.get(value));
+				}
+				else {
+					String newTag;
```

Comment:
this is an interesting caching solution - first it made me look again, but I think it's fine. 
nit: can you please extract the do-while statement to a private method: `getNextSyntheticTag`?

Reply:
I would rather keep it in this method. It is a small bit of code, and I find it more annoying than helpful to have it moved out.

other changes made.